routing/pathfind.go

```diff
@@ -608,7 +637,20 @@ func findPath(tx *bolt.Tx, graph *channeldb.ChannelGraph,
 		// edges), then we're done here and can exit the graph
 		// traversal early.
 		if bytes.Equal(bestNode.PubKeyBytes[:], targetVertex[:]) {
-			break
+			// In the event that we're sending payments to ourselves,
+			// we don't want to give up when starting at the source,
+			// thinking the shortest path is zero.
+			// But, at the same time, we want to ensure the second time
+			// that source shows up, it's considered 'destination reached.'
+			if distance[sourceVertex].dist == infinity &&
+				sourceNode.PubKeyBytes == targetNode.PubKeyBytes{
+				distance[sourceVertex] = nodeWithDist{
+                 			dist: infinity-1,
+					node: sourceNode,
+               			}
+			} else {
+				break
+			}
```

Comment:
An alternative implementation could be to temporarily add a second "virtual" `self'` node to the graph and use the normal routing algorithm to route from `self` to `self'`. I think it will make it easier to verify the correctness of the algorithm.

In any case, there are definitely some unit tests needed to cover the new functionality.

Reply:
First time using GitHub.  I'm assuming the reviewers run the unit tests, correct?

Reply:
As a developer you should always run unit tests yourself for code that you change. `pathfind_test.go` in this case. 

But what I meant was not just to run the tests, but also extend them to cover the new route to self scenario.

Also something to think about is how to indicate the desired channels to and from self. Without that, is this functionality useful for rebalancing?

Reply:
I have ran tests myself, sending to myself.  Works great.  It had some bugs, but I fixed those.  I'll have to poke around and see what you're expecting as the format of results you're expecting in pathfind_test.go.

"Also something to think about is how to indicate the desired channels to and from self. Without that, is this functionality useful for rebalancing?" 

That's *exactly* what I'm working on now.  Trying to figure out how to pass arguments from the CLI to the actual functions.  Once the arguments get to find-path, its easy.

Reply:
Figured out how to run the tests, after submitting.  The route to self works in practice, but doesn't work in the test case.
I think has something to do with the test sending nil as bandwidth hints.